### PR TITLE
Remove symlinks-to-hardlinks.ps1 hack

### LINKED ISSUE #2547: Using master version of vm in other crate on Windows
Building locally on windows works when I execute `powershell scripts\symlinks-to-hardlinks.ps1`.

But when I want to include the `rustpython-vm` crate in another crate from git (to always have the latest version), I get the errors about Lib not being found etc. Probably because the powershell script is not executed.

My current solution is to clone the repo myself, execute the powershell script and use a local path to `rustpython-vm` in my other crate. 

What do you think should be changed to make windows work without the script? I would be glad to help I someone explains why the script is needed and which other solutions could work.

### PR DESCRIPTION
**Pull request by coolreader18**
cc @jhoobergs @Pebaz, fixes #2547

### CODE REVIEW

**Comment by Pebaz** [comment]
Nice!

**Comment by coolreader18** [comment]
Could you check that this works in your environment? I'm not sure whether github actions has some special setup where symlinks actually do work.

**Comment by Pebaz** [comment]
Unfortunately, it failed to build on that branch :/

```
error: Python compile error from Lib/python_builtins/_frozen_importlib.py: Got unexpected token '.' at line 1 column 1
../../../Lib/importlib/_bootstrap.py

↑
  --> vm\src\frozen.rs:41:24
   |
41 |     ext_modules!(iter, dir = "Lib/python_builtins/");
   |                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: Python compile error from Lib/core_modules/copyreg.py: Got unexpected token '.' at line 1 column 1
../../../Lib/copyreg.py

↑
  --> vm\src\frozen.rs:46:24
   |
46 |     ext_modules!(iter, dir = "Lib/core_modules/");
   |                        ^^^^^^^^^^^^^^^^^^^^^^^^^

error: aborting due to 2 previous errors

error: could not compile `rustpython-vm`
```

For context, I believe that this is the same message as before.

**Comment by coolreader18** [comment]
@Pebaz could you pull and try again?

**Comment by jhoobergs** [comment]
@coolreader18 I added `rustpython-vm = { git="https://github.com/RustPython/RustPython.git", branch = "coolreader18/fix-windows-symlinks" }` to my Cargo.toml file, but still got the same error.

**Comment by jhoobergs** [comment]
I'm debugging a bit, I think I'm getting closer, but I'm not there yet.

File: derive/src/compile_bytecode.rs [hunk 1]
```diff
@@ -31,6 +31,47 @@ static CARGO_MANIFEST_DIR: Lazy<PathBuf> = Lazy::new(|| {
     PathBuf::from(env::var_os("CARGO_MANIFEST_DIR").expect("CARGO_MANIFEST_DIR is not present"))
 });
 
+fn is_git_symlink(path: &Path) -> bool {
+    use std::io::{BufRead, BufReader};
+    use std::process::{Command, Stdio};
+    let res = path.parent().and_then(|dir| {
+        let child = Command::new("git")
+            .arg("-C")
+            .arg(dir)
```
Comment (jhoobergs) [inline derive/src/compile_bytecode.rs:40]:
```suggestion
```

File: derive/src/compile_bytecode.rs [hunk 2]
```diff
@@ -115,9 +156,20 @@ impl CompilationSource {
...
         let mut code_map = HashMap::new();
-        let paths = fs::read_dir(&path).map_err(|err| {
-            Diagnostic::spans_error(self.span, format!("Error listing dir {:?}: {}", path, err))
-        })?;
+        let paths = fs::read_dir(path)
+            .or_else(|e| {
+                // handle git symlinks - git for windows will make it just a text file containing
+                // the target
+                if cfg!(windows) && is_git_symlink(path) {
+                    if let Ok(real_path) = fs::read_to_string(path) {
```
Comment (jhoobergs) [inline derive/src/compile_bytecode.rs:164]:
```suggestion
                    if let Ok(real_path) = fs::read_to_string(path.canonicalize().unwrap()) {
```

File: derive/src/compile_bytecode.rs [hunk 3]
```diff
@@ -148,15 +194,34 @@ impl CompilationSource {
...
+                            .ok()
+                            .unwrap_or(&path)
+                            .display()
+                    })
+                };
+                let code = compile_path(&path).or_else(|e| {
+                    if cfg!(windows) && is_git_symlink(&path) {
+                        if let Ok(real_path) = fs::read_to_string(&path) {
+                            return compile_path(real_path.trim().as_ref());
+                        }
```
Comment (jhoobergs) [inline derive/src/compile_bytecode.rs:216]:
```suggestion
                        if let Ok(real_path) = fs::read_to_string(path.canonicalize().unwrap()) {
                            let joined = path.parent().unwrap().join(real_path.trim());
                            return compile_path(&joined);
                        }
```

**Review by jhoobergs — COMMENTED** [review]
These changes make it work locally for me.

**Comment by coolreader18** [comment]
I think the `-C` is necessary, since it's not always guaranteed that you're going to be in [a subdirectory of] the git repository

File: derive/src/compile_bytecode.rs [hunk 4]
```diff
@@ -31,6 +31,47 @@ static CARGO_MANIFEST_DIR: Lazy<PathBuf> = Lazy::new(|| {
...
+            .ok()?;
+        let git_root = Command::new("git")
+            .arg("-C")
+            .arg(dir)
+            .arg("rev-parse")
+            .arg("--show-toplevel")
+            .output()
+            .ok()?
+            .stdout;
+        let git_root = Path::new(std::str::from_utf8(&git_root).unwrap().trim());
```
Comment (jhoobergs) [inline derive/src/compile_bytecode.rs:56]:
```suggestion
        let git_root = Command::new("git")
            .arg("-C")
            .arg(dir)
            .arg("rev-parse")
            .arg("--show-toplevel")
            .output()
            .ok()?
            .stdout;
        let git_root = Path::new(std::str::from_utf8(&git_root).unwrap().trim());
        let child = Command::new("git")
            .arg("-C")
            .arg(git_root)
            .arg("ls-files")
            .arg("-s")
            .stdin(Stdio::null())
            .stdout(Stdio::piped())
            .stderr(Stdio::piped())
            .spawn()
            .ok()?;
```

**Review by jhoobergs — COMMENTED** [review]
@coolreader18 Indeed, this seems to fix that problem.

File: derive/src/compile_bytecode.rs (hunk 4 above)
Comment (jhoobergs) [reply derive/src/compile_bytecode.rs:56]: The diff is a bit strange, I just used the `git_root` as the `dir`.

**Comment by coolreader18** [comment]
Thanks so much for fixing that!

File: derive/src/compile_bytecode.rs [hunk 5]
```diff
@@ -115,9 +156,20 @@ impl CompilationSource {
...
     ) -> Result<HashMap<String, FrozenModule>, Diagnostic> {
         let mut code_map = HashMap::new();
-        let paths = fs::read_dir(&path).map_err(|err| {
-            Diagnostic::spans_error(self.span, format!("Error listing dir {:?}: {}", path, err))
-        })?;
+        let paths = fs::read_dir(path)
+            .or_else(|e| {
+                // handle git symlinks - git for windows will make it just a text file containing
+                // the target
+                if cfg!(windows) && is_git_symlink(path) {
```
Comment (youknowone) [inline derive/src/compile_bytecode.rs:163]: what if the copied code doesn't have `.git`? e.g. downloading tarball or zip from https://github.com/RustPython/RustPython/archive/refs/heads/master.zip

File: derive/src/compile_bytecode.rs [hunk 6]
```diff
@@ -31,6 +31,47 @@ static CARGO_MANIFEST_DIR: Lazy<PathBuf> = Lazy::new(|| {
...
+            let link_path = line.splitn(2, '\t').nth(1).unwrap();
+            let link_path = git_root.join(link_path).canonicalize().ok()?;
+            if link_path == path {
+                return Some(true);
+            }
+        }
+        Some(false)
+    });
+    res.unwrap_or(false)
+}
```
Comment (jhoobergs) [inline derive/src/compile_bytecode.rs:73]:
```suggestion
```

File: derive/src/compile_bytecode.rs [hunk 7]
```diff
@@ -115,9 +156,20 @@ impl CompilationSource {
...
+        let paths = fs::read_dir(path)
+            .or_else(|e| {
+                // handle git symlinks - git for windows will make it just a text file containing
+                // the target
+                if cfg!(windows) && is_git_symlink(path) {
+                    if let Ok(real_path) = fs::read_to_string(path.canonicalize().unwrap()) {
+                        return fs::read_dir(real_path.trim());
+                    }
+                }
+                Err(e)
```
Comment (jhoobergs) [inline derive/src/compile_bytecode.rs:168]:
```suggestion
                // handle git symlinks - git for windows will make it just a text file containing
                // the target
                if cfg!(windows) {
                    if let Ok(real_path) = fs::read_to_string(path.canonicalize().unwrap()) {
                        let real_path = Path::new(real_path.trim());
                        if real_path.exists() {
                            return fs::read_dir(real_path);
                        } else {
                            return Err(e);
                        }
                    }
                }
                Err(e)
```

File: derive/src/compile_bytecode.rs [hunk 8]
```diff
@@ -148,15 +194,35 @@ impl CompilationSource {
...
+                    })
+                };
+                let code = compile_path(&path).or_else(|e| {
+                    if cfg!(windows) && is_git_symlink(&path) {
+                        if let Ok(real_path) = fs::read_to_string(path.canonicalize().unwrap()) {
+                            let joined = path.parent().unwrap().join(real_path.trim());
+                            return compile_path(&joined);
+                        }
+                    }
+                    Err(e)
```
Comment (jhoobergs) [inline derive/src/compile_bytecode.rs:219]:
```suggestion
                    if cfg!(windows) {
                        if let Ok(real_path) = fs::read_to_string(path.canonicalize().unwrap()) {
                            let real_path = Path::new(real_path.trim());
                            let joined = path.parent().unwrap().join(real_path);
                            if joined.exists() {
                                return compile_path(&joined);
                            } else {
                                return Err(e);
                            }
                        }
                    }
                    Err(e)
```

**Review by jhoobergs — COMMENTED** [review]
This should also work for the tarballs where a `.git` folder isn't there. All git related stuff is removed. Locally it works to build this.

File: derive/src/compile_bytecode.rs (hunk 5 above)
Comment (jhoobergs) [reply derive/src/compile_bytecode.rs:163]: Then it does indeed fail.

**Comment by Pebaz** [comment]
I have an opportunity to test this again tomorrow, can't wait to see if it works!

**Comment by Pebaz** [comment]
@coolreader18 @jhoobergs It worked!

I can run Python snippets now!

Thank you so much for your help and work on this! 🙂🙃🙂🙃

If you need me to test anything else let me know.

**Review by youknowone — APPROVED** [review]

**Pull request merged**